src/test/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandlerTests.java

```diff
@@ -120,4 +122,65 @@ public void testOutbound() {
 		assertThat(record.value()).isNull();
 	}
 
+	@Test
+	public void testOutboundWithTimestamp() {
+		ProducerFactory<Integer, String> producerFactory = new DefaultKafkaProducerFactory<>(
+				KafkaTestUtils.producerProps(embeddedKafka));
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(producerFactory);
+		KafkaProducerMessageHandler<Integer, String> handler = new KafkaProducerMessageHandler<>(template);
+		handler.setBeanFactory(mock(BeanFactory.class));
+		handler.afterPropertiesSet();
+
+		Message<?> message = MessageBuilder.withPayload("foo")
+				.setHeader(KafkaHeaders.TOPIC, topic1)
+				.setHeader(KafkaHeaders.MESSAGE_KEY, 2)
+				.setHeader(KafkaHeaders.PARTITION_ID, 1)
+				.setHeader(KafkaHeaders.TIMESTAMP, 1487694048607L)
+				.build();
+		handler.handleMessage(message);
+
+		ConsumerRecord<Integer, String> record = KafkaTestUtils.getSingleRecord(consumer, topic1);
+		assertThat(record).has(key(2));
+		assertThat(record).has(partition(1));
+		assertThat(record).has(value("foo"));
+		assertThat(record).has(timestamp(1487694048607L));
+	}
+
+	@Test
+	public void testOutboundWithTimestampExpression() {
+		ProducerFactory<Integer, String> producerFactory = new DefaultKafkaProducerFactory<>(
+				KafkaTestUtils.producerProps(embeddedKafka));
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(producerFactory);
+		KafkaProducerMessageHandler<Integer, String> handler = new KafkaProducerMessageHandler<>(template);
+		handler.setBeanFactory(mock(BeanFactory.class));
+		handler.afterPropertiesSet();
+
+		Message<?> message = MessageBuilder.withPayload("foo")
+				.setHeader(KafkaHeaders.TOPIC, topic1)
+				.setHeader(KafkaHeaders.MESSAGE_KEY, 2)
+				.setHeader(KafkaHeaders.PARTITION_ID, 1)
+				.build();
+
+		handler.setTimestampExpression(new SpelExpressionParser().parseExpression("T(Long).valueOf('1487694048633')"));
+
+		handler.handleMessage(message);
+
+		ConsumerRecord<Integer, String> record1 = KafkaTestUtils.getSingleRecord(consumer, topic1);
+		assertThat(record1).has(key(2));
+		assertThat(record1).has(partition(1));
+		assertThat(record1).has(value("foo"));
+		assertThat(record1).has(timestamp(1487694048633L));
+
+		Long currentTimeMarker = System.currentTimeMillis();
+		handler.setTimestampExpression(new SpelExpressionParser().parseExpression("T(System).currentTimeMillis()"));
```

Comment:
And this one can be replaced with the `new FunctionExpression<Message<?>>(m -> System.currentTimeMillis())`.
Not so critical, just in case to study something new yet 😄